File: tests/util-paths-project-slug.test.py
#!/usr/bin/env python3
"""project_slug() resolves the project dir Claude Code ACTUALLY created.

Regression guard for sonichi#2723: three inline slug formulas, each mapping
only `/`, silently named directories Claude Code never writes to on any
install whose path contains a space or a dot — i.e. every macOS app-bundle
install under `~/Library/Application Support/`. Nothing errored; the wrong
directory was simply created on first write and memory split across parallel
stores.

The fix is discovery-first: find the existing project dir describing the same
path (compared through a derivation-independent key) instead of predicting its
name. These tests pin that behavior, including the case a formula CANNOT get
right — a slug whose characters no current formula reproduces.
"""
import os
import sys
import tempfile
import unittest
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent / "src"))

from util_paths import (  # noqa: E402
    _slug_formula,
    project_slug,
    slug_derivation_key,
)

BUNDLE = "/Users/me/Library/Application Support/space.ag2.app/engine/sutando"
BUNDLE_TRUE_SLUG = "-Users-me-Library-Application-Support-space-ag2-app-engine-sutando"

#: What the pre-fix code computed, at all three call sites.
def _legacy_formula(path: str) -> str:
    return str(path).replace("/", "-")


class SlugDerivationKeyTest(unittest.TestCase):
    def test_divergent_derivations_of_one_path_share_a_key(self):
        """The three slugs one repo produced on the reporting machine agree."""
        variants = [
            BUNDLE_TRUE_SLUG,
            "-Users-me-Library-Application Support-space.ag2.app-engine-sutando",
            "-Users-me-Library-Application-Support-space.ag2.app-engine-sutando",
        ]
        keys = {slug_derivation_key(v) for v in variants}
        self.assertEqual(len(keys), 1, f"variants disagreed: {keys}")

    def test_unrelated_projects_do_not_collide(self):
        self.assertNotEqual(
            slug_derivation_key(BUNDLE_TRUE_SLUG),
            slug_derivation_key("-Users-me-Documents-unrelated-repo"),
        )


class SlugFormulaTest(unittest.TestCase):
    def test_legacy_formula_was_wrong_for_bundle_paths(self):
        """The bug, pinned: `/`-only mapping names a dir that never exists."""
        self.assertNotEqual(_legacy_formula(BUNDLE), BUNDLE_TRUE_SLUG)

    def test_formula_reproduces_the_bundle_slug(self):
        """Fallback is right for the install class that motivated the fix."""
        self.assertEqual(_slug_formula(BUNDLE), BUNDLE_TRUE_SLUG)

    def test_formula_is_unchanged_for_plain_paths(self):
        """Installs that already worked keep working — strict superset."""
        plain = "/Users/me/workspace/sutando"
        self.assertEqual(_slug_formula(plain), _legacy_formula(plain))


class ProjectSlugDiscoveryTest(unittest.TestCase):
    def setUp(self):
        self._tmp = tempfile.TemporaryDirectory()
        self.home = Path(self._tmp.name)
        (self.home / "projects").mkdir()
        self._prev = os.environ.get("CLAUDE_CONFIG_DIR")
        os.environ["CLAUDE_CONFIG_DIR"] = str(self.home)

    def tearDown(self):
        if self._prev is None:
            os.environ.pop("CLAUDE_CONFIG_DIR", None)
        else:
            os.environ["CLAUDE_CONFIG_DIR"] = self._prev
        self._tmp.cleanup()

    def _mkproject(self, name: str) -> Path:
        d = self.home / "projects" / name
        d.mkdir()
        return d

    def test_discovers_the_real_dir_over_the_formula(self):
        """A slug the formula does NOT reproduce is still found.

        This is the case that makes discovery worth the code: Claude Code has
        already changed how it maps `_` (one machine holds both
        `...0nw4fhvs4599_zcgpk3fbcnh0000gn...` and `...4599-zcgpk3...`), so
        the formula is a prediction that expires. Discovery does not.
        """
        real = "-Users-me-Library-Application_Support-space-ag2-app-engine-sutando"
        self._mkproject(real)
        self.assertNotEqual(_slug_formula(BUNDLE), real)  # formula cannot get here
        self.assertEqual(project_slug(BUNDLE), real)

    def test_falls_back_to_formula_when_nothing_exists(self):
        """First run: no dir yet, so any answer is a guess — use the formula."""
        self.assertEqual(project_slug(BUNDLE), _slug_formula(BUNDLE))

    def test_ignores_unrelated_projects(self):
        self._mkproject("-Users-me-Documents-unrelated-repo")
        self.assertEqual(project_slug(BUNDLE), _slug_formula(BUNDLE))

    def test_prefers_most_recently_modified_on_a_split(self):
        """The split this function exists to stop: pick the live store."""
        stale = self._mkproject(
            "-Users-me-Library-Application Support-space.ag2.app-engine-sutando"
        )
        live = self._mkproject(BUNDLE_TRUE_SLUG)
        os.utime(stale, (1_000_000, 1_000_000))
        os.utime(live, (2_000_000, 2_000_000))
        self.assertEqual(project_slug(BUNDLE), BUNDLE_TRUE_SLUG)

    def test_missing_projects_dir_is_not_an_error(self):
        for child in (self.home / "projects").iterdir():
            child.rmdir()
        (self.home / "projects").rmdir()
        self.assertEqual(project_slug(BUNDLE), _slug_formula(BUNDLE))


if __name__ == "__main__":
    unittest.main()

File: tests/util-paths-project-slug.test.ts
import { describe, it, beforeEach, afterEach } from 'node:test';
import assert from 'node:assert/strict';
import { mkdirSync, mkdtempSync, rmSync, utimesSync } from 'node:fs';
import { join } from 'node:path';
import { tmpdir } from 'node:os';
import { projectSlug, slugDerivationKey, slugFormula } from '../src/util_paths.js';

/**
 * TypeScript twin of tests/util-paths-project-slug.test.py — keep both in step.
 *
 * Regression guard for sonichi#2723: three inline slug formulas, each mapping
 * only `/`, silently named directories Claude Code never writes to on any
 * install whose path contains a space or a dot (every macOS app-bundle install
 * under `~/Library/Application Support/`). Nothing errored — the wrong dir was
 * created on first write and memory split across parallel stores.
 */

const BUNDLE = '/Users/me/Library/Application Support/space.ag2.app/engine/sutando';
const BUNDLE_TRUE_SLUG = '-Users-me-Library-Application-Support-space-ag2-app-engine-sutando';

/** What the pre-fix code computed, at all three call sites. */
const legacyFormula = (path: string) => path.replace(/\//g, '-');

describe('slugDerivationKey', () => {
	it('gives one key for the divergent derivations of a single path', () => {
		const variants = [
			BUNDLE_TRUE_SLUG,
			'-Users-me-Library-Application Support-space.ag2.app-engine-sutando',
			'-Users-me-Library-Application-Support-space.ag2.app-engine-sutando',
		];
		const keys = new Set(variants.map(slugDerivationKey));
		assert.equal(keys.size, 1, `variants disagreed: ${[...keys]}`);
	});

	it('does not collide unrelated projects', () => {
		assert.notEqual(
			slugDerivationKey(BUNDLE_TRUE_SLUG),
			slugDerivationKey('-Users-me-Documents-unrelated-repo'),
		);
	});
});

describe('slugFormula (fallback only)', () => {
	it('pins the bug: `/`-only mapping named a dir that never exists', () => {
		assert.notEqual(legacyFormula(BUNDLE), BUNDLE_TRUE_SLUG);
	});

	it('reproduces the bundle slug', () => {
		assert.equal(slugFormula(BUNDLE), BUNDLE_TRUE_SLUG);
	});

	it('is unchanged for plain paths — strict superset of the old behavior', () => {
		const plain = '/Users/me/workspace/sutando';
		assert.equal(slugFormula(plain), legacyFormula(plain));
	});
});

describe('projectSlug — discovery', () => {
	let home: string;
	let prev: string | undefined;

	beforeEach(() => {
		home = mkdtempSync(join(tmpdir(), 'sutando-project-slug-'));
		mkdirSync(join(home, 'projects'));
		prev = process.env.CLAUDE_CONFIG_DIR;
		process.env.CLAUDE_CONFIG_DIR = home;
	});

	afterEach(() => {
		if (prev === undefined) delete process.env.CLAUDE_CONFIG_DIR;
		else process.env.CLAUDE_CONFIG_DIR = prev;
		try { rmSync(home, { recursive: true, force: true }); } catch {}
	});

	const mkproject = (name: string) => {
		const d = join(home, 'projects', name);
		mkdirSync(d);
		return d;
	};

	it('finds the real dir even when the formula cannot reproduce it', () => {
		// The case that makes discovery worth the code: Claude Code has already
		// changed how it maps `_`, so any formula is a prediction that expires.
		const real = '-Users-me-Library-Application_Support-space-ag2-app-engine-sutando';
		mkproject(real);
		assert.notEqual(slugFormula(BUNDLE), real);
		assert.equal(projectSlug(BUNDLE), real);
	});

	it('falls back to the formula when nothing exists yet', () => {
		assert.equal(projectSlug(BUNDLE), slugFormula(BUNDLE));
	});

	it('ignores unrelated projects', () => {
		mkproject('-Users-me-Documents-unrelated-repo');
		assert.equal(projectSlug(BUNDLE), slugFormula(BUNDLE));
	});

	it('prefers the most recently modified on a split', () => {
		const stale = mkproject('-Users-me-Library-Application Support-space.ag2.app-engine-sutando');
		const live = mkproject(BUNDLE_TRUE_SLUG);
		utimesSync(stale, 1_000_000, 1_000_000);
		utimesSync(live, 2_000_000, 2_000_000);
		assert.equal(projectSlug(BUNDLE), BUNDLE_TRUE_SLUG);
	});

	it('treats a missing projects dir as "not yet created", not an error', () => {
		rmSync(join(home, 'projects'), { recursive: true, force: true });
		assert.equal(projectSlug(BUNDLE), slugFormula(BUNDLE));
	});
});
